### PR TITLE
webpack hot reload

### PR DESCRIPTION
**Pull request by chrisnojima**

### CODE REVIEW

File: desktop/client.js [hunk 1]
```diff
@@ -13,5 +13,12 @@ const name = 'dist/main.hot.bundle.js'
 const file = fs.createWriteStream(name)
 http.get('http://localhost:4000/dist/main.bundle.js', function (response) {
   response.pipe(file)
-  spawn(electron, [name])
+  const e = spawn(electron, [name, '--enable-logging', '--v=1'])
```
Comment (chrisnojima) [inline desktop/client.js]: extra logging and information when we debug

File: desktop/renderer/index.js [hunk 2]
```diff
@@ -20,6 +20,10 @@ let DevTools = null
 let DebugPanel = null
 let LogMonitor = null
 
+if (module.hot) {
```
Comment (chrisnojima) [inline desktop/renderer/index.js]: all hot loadable modules need this to auto refresh

File: desktop/server.js [hunk 3]
```diff
@@ -8,21 +7,21 @@ const PORT = 4000
 const compiler = webpack(config)
 
 // Just build output files and don't run a hot server
-const NO_SERVER = (process.env.NO_SERVER && process.env.NO_SERVER == 'true') || false
+const NO_SERVER = (process.env.NO_SERVER && process.env.NO_SERVER === 'true') || false
```
Comment (chrisnojima) [inline desktop/server.js]: just some lint cleanup

File: desktop/server.js [hunk 4]
```diff
@@ -42,8 +41,6 @@ if (NO_SERVER) {
 
   app.use(require('webpack-hot-middleware')(compiler))
 
-  app.get('*', (req, res) => { res.json({}) })
```
Comment (chrisnojima) [inline desktop/server.js]: no longer serving everything as {}, just 404 them

File: desktop/webpack.config.base.js [hunk 5]
```diff
@@ -46,9 +46,9 @@ module.exports = {
     __dirname: true
   },
   entry: {
-    index: './renderer/index.js',
-    main: './app/main.js',
-    launcher: './renderer/launcher.js',
-    'remote-component-loader': '../react-native/react/native/remote-component-loader.js'
+    index: ['./renderer/index.js'],
```
Comment (chrisnojima) [inline desktop/webpack.config.base.js]: this is cause we inject the hot loader in the .development config

File: desktop/webpack.config.development.js [hunk 6]
```diff
@@ -21,5 +22,14 @@ config.plugins.push(
   })
 )
 
+if (process.env.HOT === 'true') {
+  const HMR = 'webpack-hot-middleware/client?path=http://localhost:4000/__webpack_hmr'
+
+  Object.keys(config.entry).forEach(k => {
+    if (k !== 'main') { // node-only thread can't be hot loaded...
+      config.entry[k] = [HMR].concat(config.entry[k]) // Note: all entry points need `if (module.hot) {module.hot.accept()}` to allow hot auto loads to work
```
Comment (chrisnojima) [inline desktop/webpack.config.development.js]: inject the hot loader here

File: desktop/webpack.config.development.js [hunk 7]
```diff
@@ -14,6 +14,7 @@ config.output.publicPath = 'http://localhost:4000/dist/'
 // config.bail = true
 
 config.plugins.push(
+  new webpack.optimize.OccurenceOrderPlugin(),
```
Comment (chrisnojima) [inline desktop/webpack.config.development.js]: we should keep the order of the modules so hot loading isn't confused

File: react-native/react/native/notifications.js [hunk 8]
```diff
@@ -11,7 +11,7 @@ type NotifyFn = (title: string, opts: Object) => void
 
 export default function (notify: NotifyFn) {
   if (initialized) {
-    throw new Error('notifications were already initialized')
+    return
```
Comment (chrisnojima) [inline react-native/react/native/notifications.js]: if we've already inited, lets just bail and not die, else reloading won't work

**Comment by chrisnojima** [comment]
@keybase/react-hackers

**Comment by MarcoPolo** [comment]
LGTM. Does the hot reloading keep the state of the store?

**Comment by cjb** [comment]
LGTM

**Pull request merged**